### DISCUSSION TITLE
Particles injection at a specified position.

### DISCUSSION THREAD

**Discussion by rashid611** [epochpic/epoch#669] [General]
How to configure the injector scheme in EPOCH to inject particles with 50 MeV energy at the precise location within the first bubble?

![1](https://github.com/Warwick-Plasma/epoch/assets/143154854/3a99b5a2-8834-438b-aa5f-9be6144e753f)

**Comment by Status-Mirror** [comment]
Hey @rashid611, 

We have an experimental branch of EPOCH which allows you to add macro-particles to specific locations in the simulation window at specific times. It's based on the [file injector](https://epochpic.github.io/quickstart/basic_examples/injectors.html#file-injectors) method, but instead of injecting through a boundary, you specify the full injection position.

The code is currently hosted on a public EPOCH branch called [particle source](https://github.com/Warwick-Plasma/epoch/tree/particle_source), and you can access these special injectors using the `particle_source = T` command in an injector block. The remaining file-injector keys have their usual meaning.

As this is not an official release of the code, our ability to provide further assistance is limited. An example input deck would look like this:

```
begin:control
    nx = 500
    ny = 200
    t_end = 20.0e-15
    x_min = 0
    x_max = 25e-6
    y_min = -5e-6
    y_max = 5e-6
    stdout_frequency = 100
end:control

begin:boundaries
    bc_x_min = open
    bc_x_max = open
    bc_y_min = open
    bc_y_max = open
end:boundaries

begin:species
    name = Electron
    mass = 1.0
    charge = -1.0
    immobile = T
end:species

begin:injector
    species = Electron
    inject_from_file = T
    particle_source = T
    x_data = "inject_x.txt"
    y_data = "inject_y.txt"
    px_data = "inject_px.txt"
    w_data = "inject_w.txt"
    t_data = "inject_t.txt"
end:injector

begin:output
    dt_snapshot = 5.0e-15
    particles = always
end:output
```

Hope this helps,
Stuart

**Reply by rashid611** [reply]
Hi Stuart,

This will help a lot. I'll try it out and see how it works. Thank you for sharing this information and the example input deck.


Best regards,
Rashid

**Comment by rashid611** [comment]
hey @Status-Mirror ,

I have a question regarding the species blocks in the EPOCH code. If I set the species blocks as shown below with a uniform density `no=1e19`, will the resulting density be `n0, 2n0` or something else? I have neutral helium which will be ionized.

Here is the species block configuration:

```
begin:species
    name = Helium
    mass = 4.0 *1830
    charge = 0.0
    npart_per_cell=4
    immobile=T
    ionise=T
    ionisation_electron_species=(electron,electron)
    atomic_no=2
    charge = 0.0
    density = n0
end:species

begin:species
    name = Electron
    mass = 1.0
    charge = -1.0
    density = n0
    npart_per_cell=4
end:species
```

**Reply by Status-Mirror** [reply]
Hey Rashid,

EPOCH will set the weights of macro-particles in a given species to match the number density your provide it. In this case, the number density of your `Helium` species will be `n0`.

More specifically, EPOCH will look at your species block, and see 4 particles per cell with a density of `n0`. EPOCH will calculate the cell volume, $V =$ `dx * dy * dz`, where the cell-size is set to 1m in missing dimensions (e.g. `dz` = 1m in EPOCH2D). We then know the number of real particles which should be in the cell - this is $n_0 V$. If there are 4 macro-particles per cell, each one will represent $n_0 V / 4$ particles for this species.

If all your `Helium` macro-particles become ionised, then the number density of `Helium2` will be `n0`, and you'll have a number density of `2*n0` for your electrons. Quick warning though - EPOCH species names are case-sensitive. Your `ionisation_electron_species` "electron" will not be linked to your species "Electron".

Hope this helps,
Stuart

**Reply by rashid611** [reply]
Hi Stuart,

Thank you for the detailed explanation.

I now understand how EPOCH sets the weights of macro-particles to match the specified number density n0 . The process of calculating the number of real particles in a cell by considering the cell volume and determining the representation of each macro-particle is clear.

I appreciate the heads-up about the case sensitivity of species names in EPOCH. I will make sure to change "Electron" to "electron" to ensure everything is linked correctly.

Thanks again for your assistance. This information is very helpful for accurately setting up my simulations.

Best regards,
Rashid
